### DISCUSSION TITLE
How to set default advanced options in Python modules?

### DISCUSSION THREAD

**Discussion by byt3bl33d3r** [rapid7/metasploit-framework#18556] [Q&A] [answered]
Heya,

I have a few Python modules that I wrote that we'll hopefully be contributing here soon (yay), some of these require the `MeterpreterTryToFork` option to be set to true. Is there a way to automatically set this advanced option for Python Modules?

Thanks

**Comment by lvzhengcq** [comment]
In Python, you can set default options or parameters for your modules by using module-level variables or by defining functions with default arguments. Here's an example of each approach:

1.Using Module-level Variables:
You can define module-level variables at the beginning of your module and use them as defaults throughout the module. For example:

# mymodule.py

DEFAULT_OPTION_1 = 'default_value_1'
DEFAULT_OPTION_2 = 42

def my_function(option1=DEFAULT_OPTION_1, option2=DEFAULT_OPTION_2):
# Your function implementation here
pass

Users of your module can then use the default values or override them as needed.

2.Using Functions with Default Arguments:
You can also define functions with default arguments directly in your module:

# mymodule.py

def my_function(option1='default_value_1', option2=42):
# Your function implementation here
pass

In this case, users of your module can either use the default values or provide their own values when calling the function.
Remember that using module-level variables can make the code more readable and maintainable, especially when there are multiple functions using the same defaults. However, using functions with default arguments provides a more concise way to set defaults for specific functions.
When designing modules, it's essential to strike a balance between flexibility and simplicity, ensuring that users can easily understand and work with your module.

**Reply by byt3bl33d3r** [reply]
thank you chatgpt

**Comment by timwr** [comment]
I think you can just add them in the `DefaultOptions` section but I didn't check:
https://github.com/rapid7/metasploit-framework/blob/c31b0f0dd164003ac9fca3a641a6f90cb7447ed6/modules/exploits/windows/smb/cve_2020_0796_smbghost.rb#L50

**Reply by byt3bl33d3r** [reply]
I tried `default_options`, `defaultoptions` and `DefaultOptions` as the key in the metadata dictionary and they all don't seem to set the option:
```python
metadata = {
     ...
    'default_options': {
        'MeterpreterTryToFork': True
    },
```

**Comment by adfoster-r7** [comment]
Looks like this is missing functionality, I can throw up a PR to close the gap

**Reply by adfoster-r7** [reply]
Pull request created now: https://github.com/rapid7/metasploit-framework/pull/18580

If you're able to pull in those changes to your branch and verify that it works with the key ~~following example usage key of `describe_payload_options`~~ `default_options` that would help out with testing:

```python
metadata = {
    # ....
    'type': 'dos',
    'options': {
        'rhost': {'type': 'address',
		'description': 'Target address',
		'required': True,
		'default': 'None'},
        'timeout': {'type': 'int',
		'description':
		'Timeout in seconds',
		'required': True,
		'default': 15},
        'foo': {'type': 'int',
        'description':
        'Timeout in seconds',
        'required': False}
    },
    'default_options': {
        'foo': '50'
    }
}
```

~~I'll need to track down why it's called such an odd option name - https://github.com/rapid7/metasploit-framework/pull/11333/commits/0908d5a2d2e5421fffa84609329f97684836b707~~

**Reply by byt3bl33d3r** [reply]
Changes worked for me! Thanks y'all!

**Reply by adfoster-r7** [reply, marked as answer]
Thanks for testing; The code changes to support `default_options` in external Python modules has been merged now - https://github.com/rapid7/metasploit-framework/pull/18580

Will be available in the next release (6.3.46) which is triggered in ~20 minutes from now :+1:
